enclave/core/sgx/calls.c

```diff
@@ -1255,6 +1248,52 @@ void oe_abort_with_td(oe_sgx_td_t* td)
             host_ecall_context->debug_eexit_rsp = frame[0] + 8;
             host_ecall_context->debug_eexit_rip = frame[1];
         }
+
+        // For debug enclaves, log the backtrace before marking the enclave as
+        // aborted.
+        {
+            // Fetch current values of FS and GS. Typically, FS[0] == FS and
+            // GS[0] == GS.
+            uint64_t fs;
+            uint64_t gs;
+            asm volatile("mov %%fs:0, %0" : "=r"(fs));
+            asm volatile("mov %%gs:0, %0" : "=r"(gs));
+
+            // Restore FS if FS has been modified.
+            if (fs != gs)
+            {
+                // wrfsbase could trigger an exception. The enclave may not be
+                // in a state to emulate the instruction. Therefore, just
+                // restore FS[0].
+                asm volatile("mov %0, %%fs:0" : : "r"(gs) : "memory");
+            }
+
+            void* buffer[OE_BACKTRACE_MAX];
+            int size;
+            oe_result_t r = OE_UNEXPECTED;
+            if ((size = oe_backtrace(buffer, OE_BACKTRACE_MAX)) > 0)
```

Comment:
`oe_backtrace` does memory allocation though. The safer approach is checking whether the abort() is due to out of memory. Something like

`if (!out_of_memory && ((size = oe_backtrace(buffer, OE_BACKTRACE_MAX)) > 0)`

But I'm not sure how easy to decide the value if `out_of_memory`.

Reply:
Are you referring to `oe_backtrace_symbols` ? That needs memory allocatio. `oe_backtrace` shouldn't (and doesn't) allocate memory.

Reply:
Ah, yes, I was referring to the wrong function.